DataSetExplorer/AnnotationConsistencyTests/ManovaTest.cs

```diff
@@ -0,0 +1,93 @@
+﻿using CodeModel.CaDETModel.CodeItems;
+using DataSetExplorer.DataSetBuilder.Model;
+using DataSetExplorer.DataSetSerializer;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+
+namespace DataSetExplorer.AnnotationConsistencyTests
+{
+    public class ManovaTest
+    {
+        private readonly string _manovaScriptFile = "../../../AnnotationConsistencyTests/manova_test.py";
+        private readonly string _pythonPath = "../../../AnnotationConsistencyTests/venv/Scripts/python.exe";
+        private string _annotatedInstancesFile;
+        private string _metrics;
+        private string _independentVariable;
+        private readonly string _annotatedInstancesFolderPath = "D:/ccadet/annotations/sanity_check/Output/";
+
+        public delegate void TestDelegate(int id, List<DataSetInstance> instances, string codeSmell, List<CaDETMetric> metrics);
+
+        public void Run(IEnumerable<IGrouping<string, DataSetInstance>> instancesGroupedBySmells,
+            int id, TestDelegate test)
+        {
+            foreach (var codeSmellGroup in instancesGroupedBySmells)
+            {
+                var codeSmell = codeSmellGroup.Key.Replace(" ", "_");
+                var metrics = codeSmellGroup.First().MetricFeatures.Keys.ToList();
+                var instances = codeSmellGroup.ToList();
+
+                test(id, instances, codeSmell, metrics);
+            }
+        }
+
+        public void TestForMultipleAnnotators(int severity, List<DataSetInstance> instances, string codeSmell, List<CaDETMetric> metrics)
+        {
+            string exportedAnnotatorsFile = ExportAnnotatorsForSeverity(severity, instances, codeSmell);
+            SetupTestArguments(exportedAnnotatorsFile, metrics, "Annotator");
+            ProcessStartInfo startInfo = new ProcessStartInfo
+            {
+                FileName = _pythonPath,
+                Arguments = $"{_manovaScriptFile} {_annotatedInstancesFile} \"{_metrics}\" {_independentVariable}",
+                UseShellExecute = false,
+                RedirectStandardOutput = true
+            };
+            using Process process = Process.Start(startInfo);
+            using StreamReader reader = process.StandardOutput;
+            string result = reader.ReadToEnd();
+            Console.Write(result);
+        }
+
+        private string ExportAnnotatorsForSeverity(int severity, List<DataSetInstance> instances, string codeSmell)
+        {
+            var exportedAnnotatorsFile = "SanityCheck_" + codeSmell + "_Severity_" + severity;
+            var exporter = new AnnotationConsistencyByMetricsExporter(_annotatedInstancesFolderPath);
+            exporter.ExportAnnotatorsForSeverity(severity, instances, exportedAnnotatorsFile);
+            return exportedAnnotatorsFile;
+        }
+
+        public void TestForSingleAnnotator(int annotatorId, List<DataSetInstance> instances, string codeSmell, List<CaDETMetric> metrics)
```

Comment:
The problem with the current implementation is that the client of ManovaTest (Program) needs to know all the details of this class, to select the appropriate delegate. It is tightly coupled to this object. There are a few things to consider here:

- What is the difference in code between this method and `TestForMultipleAnnotators` apart from the first line and the third argument of the second line? Also examine the RunTest of AnovaTest.cs, as well as the similarities in the classes' fields.
- Maybe it would be helpful to look at the big picture - what do the classes Manova and Anova test do in a broader application - what is their purpose? Are they implementations of the same interface that holds a domain significant name (e.g., AnnotatorConsistencyChecker)?
- In such a system, what are the data flows? How can we hide the details (including which delegate to choose) from the client of these Tests? Introducing (at least imaginary) interfaces helps us understand what the role of the code implementing the interface is, and how to encapsulate all the details apart from the (often single) meaningful function call.

My advice is to think about this problem because it is a good one - sketch some class and communication diagrams of this larger system, think about the different moving parts, and how you could make the client unaware of which statistical test it is working with. Examine the easiest way to support a third type of statistical test. It takes time, but this is where your design skills improve.